### PR TITLE
Make VideoInterface aware of Double Strike modes's aspect ratio.

### PR DESCRIPTION
**Pull request by phire**
Dolphin previously didn't need to know about double strike, as it just assumed everything was 4:3.

Now we need to know that the TV will be skipping every second line. 

Actually, now that I'm writing this pull request: the code will be wrong for interlaced (non-force progressive) too. This will need more work.

Anyway, this fixes NES/SNES games after @mirrorbender's aspect ratio fixes.

### CODE REVIEW

**Comment by t27duck** [comment]
Ran through a couple NES and SNES games and also booted up Sonic Colors and Excite Truck for good measure.

My settings:
Aspect Ratio: Auto
Backend: OpenGL
Internal Resolution: Native
Anti-Aliasing: None
Post-Processing: off
Wii Aspect Ratio: 16:9
No game-specific INIs

Good news is, looks like the NES games are displaying correctly now...

fa8e01-2

However, SNES games now have the issue of instead of being squished horizontally, they are now squished vertically.

jaae01-1

jcwe01-1

Same thing with Genesis games...

mahe8p-1

While Excite Truck looked ok, Sonic Colors also exhibited the same issue.

snce8p-1

**Comment by tony971** [comment]
@t27duck http://board.byuu.org/phpbb3/viewtopic.php?f=4&t=634&sid=0dd892f325de8493ef19d2db35ac58cd&start=40#p14606

This is correct.

**Comment by t27duck** [comment]
@tony971 I'm referring to the fact that the images are too thin, not the black bars at the top and bottom. The SNES and Genesis shots should be about the same width as the Kirby picture. They are direct screen shots coming from Dolphin.

**Comment by tony971** [comment]
Ahh I thought there might have been some screenshot image distortion from github or elsewhere. Carry on.

**Comment by phire** [comment]
Ok, looks like 480p signals are being miss-interpreted as 240p signals and getting their height doubled.

**Comment by phire** [comment]
It now handles both dual strike and 480p progressive modes correctly.

**Comment by t27duck** [comment]
Kirby's Adventure: :white_check_mark: 
Super Mario Kart: :white_check_mark: 
Super Mario World: :white_check_mark: 
Sonic the Hedgehog: :white_check_mark: 
Sonic Colors: :white_check_mark: 
Excite Truck: :white_check_mark: 

Thanks for the work on this. From my testing (read: ran game for ~20 seconds just to check how it displayed) this seems like this change handles things correctly now.

:+1:

**Comment by magumagu** [comment]
The way this PR reasons about the display mode is a little weird.  There are two modes which matter: progressive and interlaced.  In progressive (480p) mode, the TV is ACV lines tall; otherwise it's 2*ACV lines tall.  The difference between "double-strike" and true interlaced mode doesn't really matter until someone implements deinterlacing.

**Comment by phire** [comment]
Yeah, this PR is kind of going about this the wrong way.

The issue is the aspect ratio code is hard-coded to one of two resolutions, 720x486 for NTSC and 720x576 for PAL.

Dual-strike mode is a true progressive mode, with a resolution of 720x243 (or 720x288 for PAL) at 60hz, which trips up the aspect ratio code, because it thinks it's seeing the middle 240 lines of a 480p signal and then dolphin renders the games with a 8:3 aspect ratio.

**Comment by waddlesplash** [comment]
*bump*
It'd be nice if the NES VC games actually displayed properly...

**Comment by JMC47** [comment]
Bump again.  I'd like this merged in soon.

File: Source/Core/Core/HW/VideoInterface.cpp [hunk 1]
```diff
@@ -436,10 +436,42 @@ u32 GetXFBAddressBottom()
...
+
+static u32 GetHalfLinesPerOddField()
+{
+	return (3 * m_VerticalTimingRegister.EQU + m_VBlankTimingOdd.PRB + 2 * m_VerticalTimingRegister.ACV + m_VBlankTimingOdd.PSB);
+}
+
+static bool IsInterlaced()
+{
+	int odd = m_VBlankTimingOdd.PRB + m_VerticalTimingRegister.EQU * 3 + m_VBlankTimingOdd.PSB;
+	int even = m_VBlankTimingEven.PRB + m_VerticalTimingRegister.EQU * 3 + m_VBlankTimingEven.PSB;
```
Comment (degasus) [inline Source/Core/Core/HW/VideoInterface.cpp]: odd and even are not used any more.

File: Source/Core/Core/HW/VideoInterface.cpp [hunk 2]
```diff
@@ -436,10 +436,42 @@ u32 GetXFBAddressBottom()
...
+	// If either field has an odd number of halflines post vertical blanking, it is an interlaced signal.
+	return GetHalfLinesPerEvenField() & 1 || GetHalfLinesPerOddField() & 1;
+}
+
+static bool IsDoubleStrike()
+{
+	int halflines = GetHalfLinesPerEvenField() + GetHalfLinesPerOddField();
+	// Normally pal has 625 lines and ntsc has 525 lines. But the double strike modes have weird timings.
+	return halflines == 624 * 2 || halflines == 526 * 2;
```
Comment (degasus) [inline Source/Core/Core/HW/VideoInterface.cpp]: I doubt we should use a black list for detecting double strike :/

Comment (phire) [inline Source/Core/Core/HW/VideoInterface.cpp]: Technically, it's a whitelist.

**Comment by phire** [comment]
The correct way to resolve this is to calculate how a real tv will act with a double strike signal. 

Dolphin currently has a special case for both NTSC (525 lines) and PAL (625 lines) and this PR in (it's current state) add special cases for both dual strike modes. What really happens (at least older non-digital TVs) the TV doesn't really care about number of lines or the timing. They simply synchronizes the movement of the electron beam to the v-sync and h-sync signals, putting the first line at the top of the screen and the last line at the bottom of the screen, 

You can have as many or as few lines per field or fields per line as you want, as long as you don't exceed certain minimum/maximum timings (such as the minimum amount of time it takes for the electron beam to get from the bottom of the screen to the top, or the right of the screen to the left, or the maximum amount of time it takes for phosphor's glow to decay). If you meet these timings (which will be different for every single TV) you will get a valid image.

**Comment by degasus** [comment]
Any updates on this PR? @phire

**Comment by JosJuice** [comment]
Is this the same as PR #3346?

**Comment by phire** [comment]
@JosJuice when this feature is implemented correctly (in whatever PR) it will get merged.

Right now #3346 is better than this attempt, but I'm still not happy, so I'm closing this one.

**Pull request closed without merging**